### PR TITLE
Implement shell quorum balancing

### PR DESCRIPTION
**Pull request by AndrejMitrovic**
Requires:
~#800~, ~#840~, ~#810~, ~#841~, ~#881~

### CODE REVIEW

**Comment by AndrejMitrovic** [comment]
Update: Added #810 to the list of required PRs. I marked that commit as `[remove]` in the PR.

File: source/agora/node/Validator.d
Comment (AndrejMitrovic) [inline source/agora/node/Validator.d]: Unrelated change.

**Comment by AndrejMitrovic** [comment]
Need to fix the integration test.

File: source/agora/consensus/Quorum.d
Comment (Geod24) [inline source/agora/consensus/Quorum.d:63]: That's how I felt

Comment (AndrejMitrovic) [reply source/agora/consensus/Quorum.d:63]: Yeah sometimes I find myself abusing ranges just to write something clever. I could have just written 2, 4, 8, 16, 32, 64, 128.. etc I guess.

**Comment by AndrejMitrovic** [comment]
It seems the test-suite times out (30 minutes). I'm trying to bisect it.

File: source/agora/test/Base.d
Comment (AndrejMitrovic) [inline source/agora/test/Base.d]: This change should be moved to the previous commit where the function was introduced.

**Comment by AndrejMitrovic** [comment]
I've discovered that what is causing these failures is actually the code coverage mode. The test-suite works fine without it. But with it there are timeouts. Either the code coverage is significantly slowing down the tests, or there is a bug in the coverage generator code in DMD front-end.

**Comment by AndrejMitrovic** [comment]
So far I can't reproduce the segfault when running github CI locally, but I can reproduce the timeout issue:

```
(lldb) bt
error: need to add support for DW_TAG_base_type 'char' encoded with DW_ATE = 0x10, bit_size = 8
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
  * frame #0: 0x00007fff6099e9de libsystem_kernel.dylib`__ulock_wait + 10
    frame #1: 0x00007fff60a5f6de libsystem_pthread.dylib`_pthread_join + 358
    frame #2: 0x000000010e81e485 agora-unittests`thread_joinAll + 197
    frame #3: 0x000000010e839225 agora-unittests`rt_term + 69
    frame #4: 0x000000010e83977c agora-unittests`_D2rt6dmain212_d_run_main2UAAamPUQgZiZ6runAllMFZv + 172
    frame #5: 0x000000010e839537 agora-unittests`_d_run_main2 + 391
    frame #6: 0x000000010e83939d agora-unittests`_d_run_main + 141
    frame #7: 0x000000010deaecd5 agora-unittests`main(argc=1, argv=0x00007ffee1ea4320) + 37
    frame #8: 0x00007fff608673d5 libdyld.dylib`start + 1
```

It seems the process is stuck on `thread_joinAll`

**Comment by AndrejMitrovic** [comment]
Ok I've added a Thread.sleep call to work around #869, but now there is a consistent segfault in the NetworkClient. I've seen this error before too:

```
Thread 40 "agora-unittests" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffe836b0700 (LWP 19915)]
0x00000000081a4493 in _D5agora7network13NetworkClientQp12sendPreimageMFNeKSQBz9consensus4data12PreImageInfoQoZ9__lambda2MFZ12__dgliteral1MFNaNiNfZv ()
    at /mnt/c/dev/agora/source/agora/network/NetworkClient.d:327
327                 this.attemptRequest(this.api.receivePreimage(preimage), null);
(gdb)

#2  0x00000000081a444d in _D5agora7network13NetworkClientQp12sendPreimageMFNeKSQBz9consensus4data12PreImageInfoQoZ9__lambda2MFZv ()
    at /mnt/c/dev/agora/source/agora/network/NetworkClient.d:327
327                 this.attemptRequest(this.api.receivePreimage(preimage), null);
(gdb) p api
$1 = (agora.api.Validator.API *) 0x0
```

So the `api` is null.

**Comment by AndrejMitrovic** [comment]
It seems that disabling parallelism fixes the issue. Unfortunately I cannot reproduce the issue locally. And also, when running the CI test-suite under LLDB it also cannot reproduce the issue.

**Comment by AndrejMitrovic** [comment]
There is something wrong with the implementation of `dispatch` in `FiberScheduler`. We lifted it from `std.concurrency`, and it was documented as an "example scheduler". 

This is my workaround which gets rid of segfaults: https://github.com/AndrejMitrovic/localrest/compare/2cf119e759cdbbe2879bf65435efd18084620498...6f35619393bfdfbb4b9f84379da68d8f40c3675a#diff-7e552a35c6e12cbbfb124195a3e45cd2

I would guess that it has something to do with manipulating the `m_fibers` array both in the `dispatch` routine and in `create()`: https://github.com/AndrejMitrovic/localrest/compare/AndrejMitrovic:2cf119e...AndrejMitrovic:6f35619#diff-7e552a35c6e12cbbfb124195a3e45cd2R936

The end result before this patch is that the fiber somehow ends up jumping to a random source code location. So it seems like the fiber gets corrupt somehow.

I still need to work on reducing this further down to a test-case..

**Comment by AndrejMitrovic** [comment]
It turned green. wow

**Comment by AndrejMitrovic** [comment]
Blocked on https://github.com/bpfkorea/agora/issues/824

And https://github.com/bpfkorea/agora/pull/889 should improve the way networking is done after the re-shuffle.

**Comment by AndrejMitrovic** [comment]
Rebased. Still blocked, and I had to revert the change Mathias made in order to fix the tests.

**Comment by AndrejMitrovic** [comment]
I'll try to unblock this by changing the test to use new validator sets instead of re-enrolling existing validators.

**Comment by AndrejMitrovic** [comment]
I should split off the first 4 commits so they can be merged right away.

**Comment by AndrejMitrovic** [comment]
Test-suite passes, but it's flakey unfortunately.

There still seems to be an issue where a task (periodic preimage reveal) starts after a sleep call and ends up segfaulting. I haven't tracked it down yet.

**Comment by AndrejMitrovic** [comment]
The feature I discovered in https://github.com/Geod24/localrest/pull/81 came in useful. I rewrote one test.

**Comment by AndrejMitrovic** [comment]
I further simplified the changeset. The diff is much smaller now.

**Comment by AndrejMitrovic** [comment]
@bpfkorea/core please review when you have the time.

File: source/agora/node/Validator.d
Comment (TrustHenry) [inline source/agora/node/Validator.d:129]: agora.consensus.Quorum.buildQuorumConfig ( const ref PublicKey node_key,
    in Enrollment[] enrolls, in UTXOFinder finder )
I think it's pointing here.
There's no problem. But The code is difficult to read.

File: source/agora/node/Ledger.d
Comment (TrustHenry) [inline source/agora/node/Ledger.d:248]: this.enroll_man.pool.remove
The `remove` is run twice here.

File: source/agora/node/Ledger.d
Comment (Geod24) [inline source/agora/node/Ledger.d:245]: https://github.com/bpfkorea/agora/blob/d2931dd0d0f552997bc80e799a5275f5c3cecff5/source/agora/consensus/EnrollmentManager.d#L214

**Review by TrustHenry — COMMENTED** [review]
Overall, it looks Great.

File: source/agora/consensus/Quorum.d
Comment (MichaelKim20) [inline source/agora/consensus/Quorum.d:96]: I think it would be nice to have a test code for `isQuorumSetSane`.

File: source/agora/consensus/Quorum.d
Comment (MichaelKim20) [inline source/agora/consensus/Quorum.d:100]: I think it would be nice to have a test code for `normalizeQSet `.

File: source/agora/test/Quorum.d
Comment (Geod24) [inline source/agora/test/Quorum.d:116]: Question about this: Should we add the ability to enroll as soon as we freeze ?
Pro:
- Genesis does it;
- It allows to potentially solve the censorship issue if paired with a rule that more validators => Higher priority;
Cons:
- Attack vector ?

**Review by Geod24 — APPROVED** [review]

**Comment by Geod24** [comment]
I don't think the double clear should block the merge.

File: source/agora/test/Quorum.d
Comment (AndrejMitrovic) [reply source/agora/test/Quorum.d:116]: Let's file an issue for this. And yes, I wanted this feature too.

File: source/agora/node/Ledger.d
Comment (AndrejMitrovic) [reply source/agora/node/Ledger.d:248]: Where do you see it?

Comment (Geod24) [reply source/agora/node/Ledger.d:248]: https://github.com/bpfkorea/agora/pull/845#discussion_r435748483

File: source/agora/consensus/Quorum.d
Comment (TrustHenry) [inline source/agora/consensus/Quorum.d:63]: // test up to 1022 nodes

```
unittest
{
    import std.algorithm;
    import std.stdio;
    import std.range;

    uint sum;
    foreach (num_nodes; iota(1, 10).map!(n => 2 ^^ n))
       sum+=num_nodes;
    writeln(sum);
}
```
Result  : 1022

Comment (AndrejMitrovic) [reply source/agora/consensus/Quorum.d:63]: It's not testing all node count configurations. That would be too expensive (especially once the algorithm is changed).

What I am testing here is this:

```
num_nodes : 2
num_nodes : 4
num_nodes : 8
num_nodes : 16
num_nodes : 32
num_nodes : 64
num_nodes : 128
num_nodes : 256
num_nodes : 512
num_nodes : 1024
```

See?

Comment (AndrejMitrovic) [reply source/agora/consensus/Quorum.d:63]: Well actually it's not inclusive as I forgot to use `11` in the iota call. +1 points to you.

File: source/agora/node/Ledger.d
Comment (AndrejMitrovic) [reply source/agora/node/Ledger.d:245]: Ah, right. I think in my fixes for #836 I will remove the call within `addValidator` because that function can fail.

**Pull request merged**
